### ISSUE TITLE
Brainstorm ideas for developer discoverability

### ISSUE THREAD

**Issue by joemasilotti** [joemasilotti/railsdevs.com#768]
It's been brought up a few times and I agree, developer discoverability can be improved.

Currently, developers are sorted by most recently added. In theory, this gives new folks a leg up over the rest of the field. In practice, it means folks get messaged a week or two after joining and then rarely ever again.

I'd like to revamp how developer search results are sorted and displayed. I'm not (yet) looking for code or implementation. But I'd love to have a brainstorm session together!

Previous ideas include: random sorting, score-based sorting, sorting by most recently _updated_ (with a cooldown), more badges, and boosting folks with more filled out profiles.

Feel free to comment below with your ideas or come talk to me IRL (or on Zoom) at the hackathon. I'll be converting this to a Discussion if it becomes hard to follow the conversation.

**Comment by RomanTurner** [comment]
Scoring Idea: Peer recommendation system. Other developers can give each other a recommendation on the platform like on Linkedin. 

Pros: 
- This could highlight specific talents for developers. 
- These could be skill tags that can show up on the developer cards. 
- Encourages developers to help each other out on the platform. 

Cons: 
- Could turn into a popularity contest.
- Doesn't properly reflect the skills of the developer.

**Comment by joemasilotti** [comment]
I love the idea of skill cards. I know businesses are already searching for specific terms (Hotwire, AWS, Tailwind, Slim, etc.) so having a specific place for folks to highlight those skills sounds great.

The two cons you listed can't be overlooked. Especially because how much results could be skewed if this becomes a popularity contest.

How does LinkedIn handle it?

**Comment by joemasilotti** [comment]
## Score-based sorting

One idea I'm been messing with is sorting based on a _score_. A score generated from a bunch of different attributes that the developer has control over.

This puts more power in the developer's hands to get bumped up the list. And it rewards folks for putting the time and effort into their profile and their presence on the platform.

Here's a first pass of the different scores that could contribute, most important/influential first.

1. **Full profile** - Boost folks who have filled out their entire profile.
2. **Response rate** - Boost folks who respond to every conversation.
3. **Source contributor** - Boost folks who have contributed to the RailsDevs codebase.
4. **Recently updated** - Boost folks who tweaked their profiles recently.
5. **Recently added** - Finally, boost newer folks on the platform.

These scores, all weighted differently, could add up to a single value which can be used to sort folks on the site. For example:

1. **Full profile** - 50 points total
    * Bio of 750+ characters - 10 points
    * 3+ social media accounts - 10 points
    * 1+ role type selected - 10 points
    * 1-2 role levels selected - 10 points
    * Availability filled in - 10 points
1. **Response rate** - 30 points total
    * 100% - 30 points
    * 90%-99% - 20 points
    * 70%-89% - 10 points
1. **Recently updated** - 15 points total
    * Updated in the last week - 15 points
    * Updated in the last 2 weeks - 10 points
    * Updated in the last month - 5 points
1. **Source contributor** - 15 points
1. **Recently added** - 5 points total
    * Added in the last week - 5 points

Keeping the thresholds and points as easily configurable options will enable fine-tuning after launch.

But I am worried that it will keep the same exact developers at the top of the list. So I think there still needs to be a significant tie-breaker. Maybe that's random?

**Comment by joemasilotti** [comment]
## Expertise bags

A lot of time and effort has been spent trying to *sort* developers optimally. But what if businesses were able to narrow down their search results so much that sorting didn’t matter?

Businesses are commonly filtering down developers by technologies, industries, and stacks. Currently, they can run a plain-text query that searches heros and bios. But it is fairly limited as it can only search for a single, exact match query.

A different approach is to offer 3-5 **expertise tags** for each developer to add to their profile. This would enable businesses to filter by one or more tags, additive.

Limiting the number of tags also ensures developers choose how they want to market themselves. It also ensures folks can’t spam the system by_all_ of them.

One potential problem is the management of this list. It would suck for a developer to want to add “AWS” but it isn't listed. Do they have to request a new tag get added? Can they add an “unofficial” tag and it doesn’t appear in filters until it becomes official?

There needs to be a easy way for a) me to add new tags and b) developers to request new ones.

This also probably includes guidelines and rules for these tags to make sure new ones aren’t arbitrarily chosen.

**Comment by joemasilotti** [comment]
## Randomization

Some ideas if we go down the randomization route... Instead of trying to "sort by random", instead run a daily/weekly job that assigns every developer a random number. Index and sort by this when displaying developers.

This ensures no random work is done when displaying results _and_ ensures the results are kept in the same order until the next task is run.
